Library/Homebrew/cmd/fetch.rb

```diff
@@ -176,81 +218,114 @@ def run
-          remaining = FETCH_MAX_TRIES - @fetch_tries[formula]
-          what = Utils.pluralize("tr", remaining, plural: "ies", singular: "y")
+                finished_downloads, remaining_downloads = remaining_downloads.partition do |_, future|
+                  finished_states.include?(future.state)
+                end
 
-          ohai "Retrying download in #{wait}s... (#{remaining} #{what} left)"
-          sleep wait
+                finished_downloads.each do |downloadable, future|
+                  previous_pending_line_count -= 1
+                  $stdout.print Tty.clear_to_end
+                  $stdout.flush
+                  output_message.call(downloadable, future)
+                end
 
-          formula.clear_cache
-          @fetch_tries[formula] += 1
-          true
-        else
-          Homebrew.failed = true
-          false
-        end
-      end
+                previous_pending_line_count = 0
+                remaining_downloads.each do |downloadable, future|
+                  # FIXME: Allow printing full terminal height.
+                  break if previous_pending_line_count >= [concurrency, (Tty.height - 1)].min
+
+                  $stdout.print Tty.clear_to_end
+                  $stdout.flush
+                  previous_pending_line_count += output_message.call(downloadable, future)
+                end
+
+                if previous_pending_line_count.positive?
+                  $stdout.print Tty.move_cursor_up_beginning(previous_pending_line_count)
+                  $stdout.flush
+                end
 
-      def fetch_fetchable(formula)
-        formula.clear_cache if args.force?
+                sleep 0.05
+              rescue Interrupt
+                remaining_downloads.each do |_, future|
+                  # FIXME: Implement cancellation of running downloads.
+                end
 
-        already_fetched = formula.cached_download.exist?
+                if previous_pending_line_count.positive?
+                  $stdout.print Tty.move_cursor_down(previous_pending_line_count - 1)
+                  $stdout.flush
+                end
 
-        begin
-          download = formula.fetch(verify_download_integrity: false)
-        rescue DownloadError
-          retry if retry_fetch?(formula)
-          raise
+                raise
+              end
+            end
+          ensure
+            $stdout.print Tty.show_cursor
+            $stdout.flush
+          end
         end
 
-        return unless download.file?
+        download_queue.shutdown
+      end
 
-        puts "Downloaded to: #{download}" unless already_fetched
-        puts "SHA256: #{download.sha256}"
+      private
+
+      def downloads
+        @downloads ||= {}
+      end
 
-        formula.verify_download_integrity(download)
+      def fetch_downloadable(downloadable)
+        downloads[downloadable] ||= download_queue.enqueue(RetryableDownload.new(downloadable), force: args.force?)
```

Comment:
Does this obsolete `--retry` or should it be conditional on that flag?

Reply:
It should be conditional on the flag, I think I just missed passing the `tries` parameter here.

Reply:
```suggestion
        tries = args.retry? ? {} : { tries: 1 }
        downloads[downloadable] ||= download_queue.enqueue(RetryableDownload.new(downloadable, **tries), force: args.force?)
```